File: java/src/de/raulin/rosario/dfa/DFA.java
package de.raulin.rosario.dfa;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class DFA {
	private int states;
	private int initial;
	private char[] alphabet;
	private Set<Integer> finalStates;
	private Map<Pair<Integer, Character>, Integer> transitions;

	public DFA(int states, int initial, String alphabet) {
		this.states = states;
		this.initial = initial;
		this.finalStates = new HashSet<Integer>();
		this.transitions = new HashMap<Pair<Integer, Character>, Integer>();
		this.alphabet = alphabet.toCharArray();
	}

	public void addTransition(Integer from, Character input, Integer to) {
		transitions.put(new Pair<Integer, Character>(from, input), to);
	}

	public void addFinalState(Integer state) {
		finalStates.add(state);
	}

	public boolean accepts(String input) {
		Integer state = initial;

		for (int i = 0; i < input.length(); ++i) {
			Pair<Integer, Character> key = new Pair<Integer, Character>(state,
					input.charAt(i));
			if (transitions.containsKey(key)) {
				state = transitions.get(key);
			} else {
				return false;
			}
		}

		return finalStates.contains(state);
	}

	public Map<Integer, List<Integer>> getEquivClasses() {
		boolean[][] table = new boolean[states][states];

		for (Integer state : finalStates) {
			for (int i = 0; i < states; ++i) {
				if (!finalStates.contains(i)) {
					table[state][i] = true;
					table[i][state] = true;
				}
			}
		}

		boolean change;
		do {
			change = false;

			for (int i = 0; i < states; ++i) {
				for (int j = 1; j < i; ++j) {
					if (i !=j && table[i][j] == false) {
						for (int k = 0; k < alphabet.length; ++k) {
							Integer t1 = transitions
									.get(new Pair<Integer, Character>(i,
											alphabet[k]));
							Integer t2 = transitions
									.get(new Pair<Integer, Character>(j,
											alphabet[k]));

							if (table[t1][t2] == true) {
								change = true;
								table[i][j] = true;
								table[j][i] = true;
							}
						}
					}
				}
			}
		} while (change == true);

		Map<Integer, List<Integer>> classes = new HashMap<Integer, List<Integer>>();
		for (int i = 0; i < states; ++i) {
			boolean newClass = true;
			for (int j = 0; j < i; ++j) {
				if (table[i][j] == false) {
					if (classes.get(j) == null) {
						List<Integer> li = new ArrayList<Integer>();
						li.add(i);
						classes.put(j, li);
					} else {
						classes.get(j).add(i);
					}
					newClass = false;
				}
			}
			if (newClass == true) {
				List<Integer> li = new ArrayList<Integer>();
				li.add(i);
				classes.put(i, li);
			}
		}

		return classes;
	}
}

File: DFA.cpp
//
//  DFA.cpp
//  dfa
//
//  Created by Rosario Raulin on 24.07.13.
//  Copyright (c) 2013 Rosario Raulin. All rights reserved.
//

#include "DFA.hpp"

DFA::DFA(const int& start, const int& states, const std::string& alphabet) :
_start(start), _states(states), _alphabet(alphabet) {}

void
DFA::addFinalState(int x) {
    _fstates.insert(x);
}

void
DFA::addTransition(int from, char input, int to) {
    _transitions[std::make_pair(from, input)] = to;
}

bool
DFA::accepts(const std::string &input) const {
    int state = _start;
    
    for (auto it = input.cbegin(); it != input.cend(); ++it) {
        auto trans = _transitions.find(std::make_pair(state, *it));
        if (trans == _transitions.end()) {
            return false;
        } else {
            state = trans->second;
        }
    }
    
    return _fstates.find(state) != _fstates.end();
}

/**
    Helper function that does the actual computation using a table filling
    algorithm.
 **/
void
DFA::fillTable(bool **table) const {
    for (auto it = _fstates.cbegin(); it != _fstates.cend(); ++it) {
        for (int i = 0; i < _states; ++i) {
            if (_fstates.find(i) == _fstates.end()) { // i is not a final state
                table[*it][i] = true;
                table[i][*it] = true;
            }
        }
    }
    
    bool change;
    do {
        change = false;
        
        for (int i = 0; i < _states; ++i) {
            for (int j = 1; j < _states; ++j) {
                if (i != j && !table[i][j]) {
                    for (auto alpha = _alphabet.cbegin();
                         alpha != _alphabet.cend();
                         ++alpha) {
                        auto d1 = _transitions.find(std::make_pair(i, *alpha));
                        auto d2 = _transitions.find(std::make_pair(j, *alpha));
                        
                        int dx = d1->second;
                        int dy = d2->second;
                        
                        if (table[dx][dy] == true) {
                            table[i][j] = true;
                            table[j][i] = true;
                            change = true;
                        }
                    }
                }
            }
        }
    } while (change == true);
}

/**
    Returns a map from a state to all equivalent states.
 **/
std::map<int, std::vector<int>>
DFA::getEquivClasses() const {
    bool** table = new bool*[_states];
    
    for (int i = 0; i < _states; ++i) {
        table[i] = new bool[_states];
        for (int j = 0; j < _states; ++j) {
            table[i][j] = false;
        }
    }
    
    fillTable(table);
    
    std::map<int, std::vector<int>> classes;
    for (int i = 0; i < _states; ++i) {
        bool isMin = true;
        for (int j = 0; j < i; ++j) {
            if (table[i][j] == false) {
                classes[j].push_back(i);
                isMin = false;
            }
        }
        if (isMin) {
            classes[i].push_back(i);
        }
        delete[] table[i];
    }
    
    delete[] table;
    
    return classes;
}